File: src/declaredObjects.py
import re


class TDNTtamplates:
    tables: dict
    Matrices: dict
    classes: list
    classNames: list


def createVar(type, templates):
    if type == "entier":
        return "0"
    elif type in ["chaine", "caractere"]:
        return '""'
    elif type == "reel":
        return "0.0"
    elif type == "booleen":
        return "False"
    elif type in templates.classes:
        return f'{type} + "()"'
    elif type in list(templates.tables.keys()):
        return createTable(
            type, templates.tables[type][0], templates.tables[type][1], templates
        )
    elif type in list(templates.Matrices.keys()):
        return createMatrix(
            type, templates.tables[type][0], templates.tables[type][1], templates
        )
    return type


def createClass(enregistrement, template):
    classlist = []
    classlist.append(f"class {list(enregistrement.keys())[0]}:\n")
    enrdesc = dict(list(enregistrement.values())[0])
    for i, v in enumerate(enrdesc):
        classlist.append(f"\t {v} = {createVar(enrdesc[v],template)} \n")
    classlist.append("\n")
    return classlist


def createTable(length, type, templates):
    tab = []
    if type == "entier": 
        return f"{str([0]*length)}"
    elif type in ["chaine", "caractere"]: 
        return f"{str([""]*length)}"
    elif type == "reel": 
        return f"{str([0.0]*length)}"
    elif type == "booleen": 
        return f"{str([False]*length)}"
    elif type in list(templates.tables.values()):
        return f"{str([]*length)}"
    elif type in templates.classes:
        return f"{str((type + '()')*length)}"
    return f"[{','.join(tab)}]\n"


def createMatrix(row, col, type, templates):
    if type == "entier":
        return f"np.array([([int()]*{col})]*{row})"
    elif type in ["chaine", "caractere"]:
        return f"np.array([([str()]*{col})]*{row})"
    elif type == "reel":
        return f"np.array([([float()]*{col})]*{row})"
    elif type == "booleen":
        return f"np.array([([bool()]*{col})]*{row})"
    elif type in templates.classNames:
        return f"np.array([([{type}()]*{col})]*{row})"
    elif type in list(templates.tables.keys()):
        return f"np.array([([{createTable(templates.tables[type][0],templates.tables[type][1],templates)}]*{col})]*{row})"
    else:
        return f"np.array([([{type}]*{col})]*{row})"


def formatTDNT(tdnt):
    newObjects = TDNTtamplates()
    newObjects.tables = {}
    newObjects.Matrices = {}
    newObjects.classes = []
    newObjects.classNames = []
    for i, V in enumerate(tdnt):
        k = list(V.keys())[0]
        v = list(V.values())[0].strip()
        if v in ["fin", "debut", "", "__SKIP__"]:
            continue
        if (v.find(":") == -1) and (v[0:8] != "tableau "):
            estr = f"declaration d'objet invalide dans TDNT ligne {i} :{v} "
            raise Exception(estr)

        key = v.strip().split(":")[0].strip()
        Ivalue = v.strip().split(":")[1].strip()
        s1 = re.match(
            "^tableau[ ]+de[ ]+(?P<long>[0-9]+)+[ ]+(?P<type>[0-9a-z]+)$",
            Ivalue.strip(),
            re.IGNORECASE,
        )
        s2 = re.match(
            "^(tableau|matrice)[ ]+de[ ]+(((?P<rows1>[0-9]+)[ ]+lignes \* (?P<cols1>[0-9]+)[ ]+colon?nes)|((?P<cols2>[0-9]+)[ ]+colon?nes) \* (?P<rows2>[0-9]+)[ ]+lignes)[ ]*(?P<type>[a-z][0-9a-z]*)$",
            Ivalue.strip(),
            re.IGNORECASE,
        )
        if Ivalue == "enregistrement":
            for j in range(i, len(tdnt)):
                if list(tdnt[j].values())[0].strip() == "fin":
                    details = tdnt[i : j + 1][2:-1]
                    for exp in range(i, j + 1):
                        tdnt[exp] = {k: "__SKIP__"}
                    detailsdict = {}
                    for H in details:
                        if H in ["fin", "debut", ""]:
                            continue

                        detailsdict[list(H.values())[0].split(":")[0]] = list(
                            H.values()
                        )[0].split(":")[1]
                    newObjects.classes.append(
                        createClass({key: detailsdict}, newObjects)
                    )
                    newObjects.classNames.append(key.strip())

                    break

        elif s1:
            if s1 and s1.group("long") and s1.group("type"):
                newObjects.tables[key] = [int(s1.group("long")), s1.group("type")]
        elif s2:
            if s2:
                if s2.group("rows1") and s2.group("cols1") and s2.group("type"):
                    newObjects.Matrices[key] = [
                        int(s2.group("rows1")),
                        int(s2.group("cols1")),
                        s2.group("type"),
                    ]
                elif s2.group("rows2") and s2.group("cols2") and s2.group("type"):
                    newObjects.Matrices[key] = [
                        int(s2.group("rows2")),
                        int(s2.group("cols2")),
                        s2.group("type"),
                    ]
    return newObjects


def formatTDO(tdo, tdnt):
    template = formatTDNT(tdnt)
    tdoDict = {}
    for i, v in enumerate(tdo):
        v = list(v.values())[0].strip()
        if v == "":
            continue
        keys = v.split(":")[0].strip()
        val = v.split(":")[1].strip()
        for j in keys.split(","):
            tdoDict[j] = val
    tdoResult = []
    for i in tdoDict:
        s1 = re.match(
            "^tableau[ ]+de[ ]+(?P<long>[0-9]+)+[ ]+(?P<type>[0-9a-z]+)$",
            tdoDict[i],
            re.IGNORECASE,
        )
        s2 = re.match(
            "^(tableau|matrice)[ ]+de[ ]+(((?P<rows1>[0-9]+)[ ]+lignes \* (?P<cols1>[0-9]+)[ ]+colon?nes)|((?P<cols2>[0-9]+)[ ]+colon?nes) \* (?P<rows2>[0-9]+)[ ]+lignes)[ ]*(?P<type>[a-z][0-9a-z]*)$",
            tdoDict[i].strip(),
            re.IGNORECASE,
        )
        if tdoDict[i] in list(template.tables.keys()):
            tbl = createTable(
                int(template.tables[tdoDict[i]][0]),
                template.tables[tdoDict[i]][1],
                template,
            )
            tdoResult.append(f"{i} = " + tbl)
        elif tdoDict[i] in list(template.Matrices.keys()):
            tbl = createMatrix(
                int(template.Matrices[tdoDict[i]][0]),
                int(template.Matrices[tdoDict[i]][1]),
                template.Matrices[tdoDict[i]][2],
                template,
            )
            tdoResult.append(f"{i} = " + tbl)
        elif s1:
            if s1 and s1.group("long") and s1.group("type"):
                tbl = createTable(int(s1.group("long")), s1.group("type"))
                tdoResult.append(f"{i} = " + tbl)
        elif s2:
            if s2:
                if (s2.group("rows1") and s2.group("cols1")) and s2.group("type"):
                    tbl = createMatrix(
                        int(s2.group("rows1")),
                        int(s2.group("cols1")),
                        s2.group("type"),
                        template,
                    )
                    tdoResult.append(f"{i} = " + tbl)
                elif (s2.group("rows2") and s2.group("cols2")) and s2.group("type"):
                    tbl = createMatrix(
                        int(s2.group("rows2")),
                        int(s2.group("cols2")),
                        s2.group("type"),
                        template,
                    )
                    tdoResult.append(f"{i} = " + tbl)
        else:
            if createVar(tdoDict[i].strip(), template):
                tdoResult.append(
                    f"{i.strip()} = " + createVar(tdoDict[i].strip(), template) + "\n"
                )
            else:
                tdoResult.append(i + f"={tdoDict[i]}\n")

    return tdoResult

File: main/IDE/mods/declaredObjects.js

class TDNTtamplates {
    tables = {}
    Matrices = {}
    classes = []
    classNames = []
}


//TODO : ADD CASCADING CHECK FOR NEW TYPES (TDNT) before executing


function createVar(type, templates) {
    if (type == "entier") { return "0" }
    else if (["chaine", "caractere"].includes(type)) { return '""' }
    else if (type == "reel") { return "0.0" }
    else if (type == "booleen") { return "False" }
    else if (templates.classNames.includes(type)) { return `new ${type} + "()"` }
    else if (Object.keys(templates.tables).includes(type)) { return createTable(type, templates.tables[type][0], templates.tables[type][1], templates) }
    else if (Object.keys(templates.Matrices).includes(type)) { return createMatrix(type, templates.tables[type][0], templates.tables[type][1], templates) }
    return type
}

function createClass(enregistrement, template) {
    let classlist = []
    let enrdesc = Object(Object.values(enregistrement)[0])
    for (const [i, v] of Object.entries(enrdesc)) { classlist.push(`${v} = ${createVar(enrdesc[v], template)}`) }
    classlist = classlist.join(';')
        `class ${Object.keys(enregistrement)[0]} {` + classlist.join(';') + "}"
    return classlist
}

function createArray(item, length) { let tea = []; for (let i = 0; i < length; i++) { tea.push(item) } return tea }

function createTable(length, type, templates) {
    if (type == "entier") { return `[${String(createArray(0, length))}]` }
    else if (type == "reel") { return `[${String(createArray(0.0, length))}]` }
    else if (type == "booleen") { return `[${String(createArray(false, length))}]` }
    else if (templates.classNames.includes(type)) { return `[${String(createArray('new ' + type + '()', length))}]` }
    else if (Object.keys(templates.Matrices).includes(type)) { return `[${String(createMatrix(templates.Matrices[key][0], templates.Matrices[key][1], templates.Matrices[key][2], templates))}]` }
    else if (Object.keys(templates.tables).includes(type)) { return `[${String(createArray(createTable(templates.tables[type][0], templates.tables[type][1], templates), length))}]` }
    else { return `[${String(createArray("", length))}]` }

}
function createMatrix(row, col, type, templates) {
    if (type == "entier") { return String((createArray(createArray(0, col), row))) }
    else if (type == "reel") { return String((createArray(createArray(0.0, col), row))) }
    else if (type == "booleen") { return String((createArray(createArray(false, col), row))) }
    else if (templates.classNames.includes(type)) { return String((createArray(createArray(`new ${templates.classNames[type]}()`, col), row))) }
    else if (Object.keys(templates.tables).includes(type)) { return String((createArray(createArray(String(createTable(templates.tables[type][0], templates.tables[type][1], templates)), col), row))) }
    else if (Object.keys(templates.Matrices).includes(type)) { return String((createArray(createArray(String(createMatrix(templates.Matrices[type][0], templates.Matrices[type][1], templates.Matrices[type][2], templates)), col), row))) }
    return createArray(createArray('', col), row)
}


function formatTDNT(tdnt) {
    let newObjects = new TDNTtamplates()
    for (const [i, V] of Object.entries(tdnt)) {
        let k = Object.keys(V)[0].trim()
        let v = Object.values(V)[0].trim()

        if (["fin", "debut", "", "__SKIP__"].includes(v)) { continue }
        if ((v.indexOf(":") == -1) && (v.substring(0, 8) != "tableau ")) {
            let estr = `declaration d'objet invalide dans TDNT ligne ${i} :${v}`
            throw estr
        }
        let key = v.split(":")[0].trim()
        let Ivalue = v.split(":")[1].trim()
        let s1 = Ivalue.match(/^tableau[ ]+de[ ]+(?<long>[0-9]+)+[ ]+(?<type>[0-9a-z]+)$/i)
        let s2 = Ivalue.match(/^(tableau|matrice)[ ]+de[ ]+(((?<rows1>[0-9]+)[ ]+lignes \* (?<cols1>[0-9]+)[ ]+colon?nes)|((?<cols2>[0-9]+)[ ]+colon?nes) \* (?<rows2>[0-9]+)[ ]+lignes)[ ]*(?<type>[a-z][0-9a-z]*)$/i)
        if (Ivalue == "enregistrement") {
            for (let j = i; j < tdnt.length; j++) {
                if (Object.values(tdnt[j])[0].trim() == "fin") {
                    let details = tdnt.slice(i, j + 1).slice(2, -1)
                    for (let exp = 0; exp < (j + 1); exp++) {
                        tdnt[exp] = { k: '__SKIP__' };
                    }
                    let detailsdict = {}
                    for (H of details) {
                        if (["fin", "debut", ""].includes(H)) { continue }
                        detailsdict[Object.values(H)[0].split(":")[0]] = Object.values(H)[0].split(":")[1]
                    }
                    let TEMP = {}; TEMP[key] = detailsdict
                    newObjects.classes.push(createClass(TEMP, newObjects)) // * classes should be for templates createClass should be called in TDO
                    newObjects.classNames.push(key)
                    break
                }
            }
        } else if (s1 && s1.groups["long"] && s1.groups["type"]) {
            newObjects.tables[key] = [parseInt(s1.groups["long"]), s1.groups["type"]]
        } else if (s2 && s2.groups["rows1"] && s2.groups["cols1"] && s2.groups["type"]) {
            newObjects.Matrices[key] = [parseInt(s2.groups["rows1"]), parseInt(s2.groups["cols1"]), s2.groups["type"],]
        } else if (s2 && s2.groups["rows2"] && s2.groups["cols2"] && s2.groups["type"]) {
            newObjects.Matrices[key] = [parseInt(s2.groups["rows2"]), parseInt(s2.groups["cols2"]), s2.groups["type"],]
        }
    }

    return newObjects
}
function formatTDO(tdo, tdnt) {
    let template = formatTDNT(tdnt)
    let tdoDict = {}
    for (let [i, v] of Object.entries(tdo)) {
        v = Object.values(v)[0]
        if (v.length === 0) { continue }
        let keys = v.split(":")[0].trim()
        let val = v.split(":")[1].trim()
        for (let j of keys.split(",")) { tdoDict[j] = val }
    }
    let tdoResult = []
    for (const i of Object.keys(tdoDict)) {
        let s1 = tdoDict[i].trim().match(/^tableau[ ]+de[ ]+(?<long>[0-9]+)+[ ]+(?<type>[0-9a-z]+)$/i)
        let s2 = tdoDict[i].trim().match(/^(tableau|matrice)[ ]+de[ ]+(((?<rows1>[0-9]+)[ ]+lignes \* (?<cols1>[0-9]+)[ ]+colon?nes)|((?<cols2>[0-9]+)[ ]+colon?nes) \* (?<rows2>[0-9]+)[ ]+lignes)[ ]*(?<type>[a-z][0-9a-z]*)$/i)
        if (Object.keys(template.tables).includes(tdoDict[i])) {
            tdoResult.push(`${i} = ` + createTable(parseInt(template.tables[tdoDict[i]][0]), template.tables[tdoDict[i]][1], template))
        } else if (Object.keys(template.Matrices).includes(tdoDict[i])) {
            tdoResult.push(`${i} = ` + createMatrix(parseInt(template.Matrices[tdoDict[i]][0]), parseInt(template.Matrices[tdoDict[i]][1]), template.Matrices[tdoDict[i]][2], template))
        } else if (s1 && s1.groups["long"] && s1.groups["type"]) {
            tdoResult.push(`${i} = ` + createTable(parseInt(s1.groups["long"].trim()), s1.groups["type"].trim()))
        } else if (s2 && s2.groups["rows1"] && s2.groups["cols1"] && s2.groups["type"]) {
            tdoResult.push(`${i} = ` + createMatrix(parseInt(s2.groups["rows1"].trim()), parseInt(s2.groups["cols1"].trim()), s2.groups["type"], template))
        } else if (s2 && s2.groups["rows2"] && s2.groups["cols2"] && s2.groups["type"]) {
            tdoResult.push(`${i} = ` + createMatrix(parseInt(s2.groups["rows2"].trim()), parseInt(s2.groups["cols2"].trim()), s2.groups["type"].trim(), template))
        } else {
            if (createVar(tdoDict[i].trim(), template)) { tdoResult.push(`${i.trim()} = ` + createVar(tdoDict[i].trim(), template)) }
            else { tdoResult.push(i + `= ${tdoDict[i]}`) }
        }
    }
    for (let i in tdoResult) {
        tdoResult[i] = tdoResult[i] + ';'
    }
    return tdoResult
}